src/systems/systems/src/pressurization/cabin_pressure_controller.rs

```diff
@@ -0,0 +1,455 @@
+        } else {
+            Ratio::new::<percent>((pressure_ratio + vs_ratio) * 100.)
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum PressureSchedule {
+    Ground,
+    TakeOff,
+    ClimbInternal,
+    Cruise,
+    DescentInternal,
+    Abort,
+}
+
+#[derive(Copy, Clone)]
+struct PressureScheduleManager {
+    pressure_schedule: PressureSchedule,
+    timer: Duration,
+    cpc_switch_reset: bool,
+}
+
+impl PressureScheduleManager {
+    fn new() -> Self {
+        Self {
+            pressure_schedule: PressureSchedule::Ground,
+            timer: Duration::from_secs_f64(100.),
+            cpc_switch_reset: false,
+        }
+    }
+
+    fn update(&mut self, context: &UpdateContext, engines: [&impl EngineCorrectedN1; 2]) {
+        match self.pressure_schedule {
+            PressureSchedule::Ground => self.update_from_ground(context, engines),
+            PressureSchedule::TakeOff => self.update_from_takeoff(context, engines),
+            PressureSchedule::ClimbInternal => self.update_from_climb(context),
+            PressureSchedule::Cruise => self.update_from_cruise(context),
+            PressureSchedule::DescentInternal => self.update_from_descent(context),
+            PressureSchedule::Abort => self.update_from_abort(context),
+        }
+    }
+
+    fn pressure_schedule(&self) -> PressureSchedule {
+        self.pressure_schedule
+    }
+
+    fn update_from_ground(
+        &mut self,
+        context: &UpdateContext,
+        engines: [&impl EngineCorrectedN1; 2],
+    ) {
+        if engines
+            .iter()
+            .all(|&x| x.corrected_n1() > Ratio::new::<percent>(70.))
+            && context.is_on_ground()
+        {
+            self.pressure_schedule = PressureSchedule::TakeOff;
+            self.timer = Duration::from_secs_f64(0.);
+            self.cpc_switch_reset = true;
+        } else if !context.is_on_ground()
+            && context.indicated_airspeed().get::<knot>() > 100.
+            && context.ambient_pressure().get::<hectopascal>() > 0.
+        {
+            self.pressure_schedule = PressureSchedule::ClimbInternal;
+            self.timer = Duration::from_secs_f64(0.);
+            self.cpc_switch_reset = true;
+        } else if self.timer < (Duration::from_secs_f64(100.)) {
+            self.timer += context.delta();
+        }
+    }
+
+    fn update_from_takeoff(
+        &mut self,
+        context: &UpdateContext,
+        engines: [&impl EngineCorrectedN1; 2],
+    ) {
+        if engines
+            .iter()
+            .all(|&x| x.corrected_n1() < Ratio::new::<percent>(70.))
+            && context.is_on_ground()
+        {
+            self.pressure_schedule = PressureSchedule::Ground;
+        } else if !context.is_on_ground()
+            && context.indicated_airspeed().get::<knot>() > 100.
+            && context.ambient_pressure().get::<hectopascal>() > 0.
+        {
+            self.pressure_schedule = PressureSchedule::ClimbInternal;
```

Comment:
**question**: Is there a specific reason why here there timer doesn't reset to 0?

Reply:
Not applicable after refactor 😄